pkg/discovery/discover_test.go

```diff
@@ -0,0 +1,548 @@
+}
+
+func (m *mockAnypointClient) GetAnalyticsWindow() ([]anypoint.AnalyticsEvent, error) {
+	return nil, nil
+}
+
+func getAgent() Agent {
+	mac := &mockAnypointClient{}
+	assetCache := cache.New()
+	buffer := 5
+
+	return Agent{
+		discoveryPageSize: 1,
+		anypointClient:    mac,
+		stage:             "Sandbox",
+		assetCache:        assetCache,
+		stopDiscovery:     make(chan bool),
+		pollInterval:      time.Second,
+		apiChan:           make(chan *ServiceDetail, buffer),
+	}
+}
+
+func TestDiscoverAPIs(t *testing.T) {
+	a := getAgent()
+
+	go func() {
+		a.discoverAPIs()
+	}()
+
+	select {
+	case sd := <-a.apiChan:
+		assert.Equal(t, sd.Stage, a.stage)
+		assert.Equal(t, sd.ResourceType, "oas2")
+		assert.Equal(t, sd.ID, "6789")
+		assert.Equal(t, sd.APISpec, []byte("{\"basePath\":\"google.com\",\"host\":\"\",\"schemes\":[\"\"],\"swagger\":\"2.0\"}"))
+
+	case <-time.After(time.Second * 1):
+		t.Errorf("Timed out waiting for the discovery")
+	}
+}
+
+type mockClient struct {
+	reqs map[string]*api.Response
+}
+
+func (mc *mockClient) Send(request api.Request) (*api.Response, error) {
+	req, ok := mc.reqs[request.URL]
+	if ok {
+		return req, nil
+	} else {
+		return nil, fmt.Errorf("no request found for %s", request.URL)
+	}
+}
+
+var mapOfResponses = map[string]*api.Response{
+	"/accounts/login": {
+		Code:    200,
+		Body:    []byte("{\"access_token\":\"abc123\"}"),
+		Headers: nil,
+	},
+	"/accounts/api/me": {
+		Code: 200,
+		Body: []byte(`{
+					"user":{
+						"identityType": "idtype",
+						"id": "123",
+						"username": "name",
+						"firstName": "first",
+						"lastName": "last",
+						"email": "email",
+						"organization": {
+							"id": "333",
+							"name": "org1",
+							"domain": "abc.com"
+						}
+					}
+				}`),
+	},
+	"/accounts/api/organizations/333/environments": {
+		Code: 200,
+		Body: []byte(`{
+					"data": [{
+						"id": "111",
+						"name": "Sandbox",
+						"organizationId": "333",
+						"type": "fake",
+						"clientId": "abc123"
+					}],
+					"total": 1
+				}`),
+	},
+}
+
+func buildAgentWithCustomMockAnypointClient() (*Agent, *mockClient) {
+	ac := &config.AgentConfig{
+		CentralConfig: corecfg.NewCentralConfig(corecfg.TraceabilityAgent),
+		MulesoftConfig: &config.MulesoftConfig{
+			PollInterval: 1 * time.Second,
+		},
+	}
+
+	mc := &mockClient{}
+	mc.reqs = make(map[string]*api.Response)
+	for k, v := range mapOfResponses {
+		mc.reqs[k] = v
+	}
+
+	apClient := anypoint.NewClient(ac.MulesoftConfig, anypoint.SetClient(mc))
+	agent := New(ac, apClient)
+
+	return agent, mc
+}
+
+func TestGetServiceDetailFailsWhenGetPoliciesFails(t *testing.T) {
+	agent, mc := buildAgentWithCustomMockAnypointClient()
+
+	mc.reqs["/apimanager/api/v1/organizations/333/environments/111/apis/456/policies"] = &api.Response{
+		Code: 400,
+		Body: nil,
+	}
+
+	asset := &anypoint.Asset{}
+
+	a := &anypoint.API{
+		ID:          456,
+		EndpointURI: "google.com",
+	}
+
+	_, err := agent.getServiceDetail(asset, a)
```

Comment:
@sagarkal Should we assert on the serviceDetail that is returned?

Reply:
serviceDetail will be nil in this test because the call to get policies fails